src/suggestion.rs

```diff
@@ -113,23 +324,75 @@ impl<'s> fmt::Display for Suggestion<'s> {
             ))
             .fmt(formatter)?;
 
-        writeln!(formatter, " {}", self.chunk.as_str())?;
-
         // underline the relevant part with ^^^^^
-        let marker_size = if self.span.end.line == self.span.start.line {
-            // column bounds are inclusive, so for a correct length we need to add + 1
-            self.span.end.column.saturating_sub(self.span.start.column) + 1
-        } else {
+
+        // @todo this needs some more thought once multiline comments pop up
+        let marker_size = self.span.one_line_len().unwrap_or_else(|| {
             self.chunk
-                .as_str()
-                .chars()
-                .count()
+                .len_in_chars()
                 .saturating_sub(self.span.start.column)
+        });
+
+        // assumes the _mistake_ is within one line
+        // if not we chop it down to the first line
+        let (line_range, start_of_line_offset) = self
+            .chunk
+            .as_str()
+            .chars()
+            .enumerate()
+            .scan(0usize, |last_newline_plus_1, (idx, c)| {
+                match c {
+                    '\n' => {
+                        *last_newline_plus_1 = idx + 1;
+                    }
+                    _ => {}
+                }
+                Some((idx, c, *last_newline_plus_1))
+            })
+            .skip_while(|(idx, _c, _last_newline_plus_1)| *idx < self.range.start)
+            .take_while(|(_idx, c, _last_newline_plus_1)| *c != '\n')
+            .last() // the last contains the last chars index, and the index of the previous newline
+            .map(|(idx, _, last_newline_plus_1)| {
+                let mistake_to_start_of_line_offset = self.range.start - last_newline_plus_1;
+                (
+                    last_newline_plus_1..(idx + 1),
+                    mistake_to_start_of_line_offset,
+                )
+            })
+            .expect("Must have at least one relevant line");
+
+        let intra_line_mistake_range = Range {
+            start: start_of_line_offset,
+            end: cmp::min(start_of_line_offset + self.range.len(), line_range.len()),
         };
+        let relevant_line = self
+            .chunk
+            .as_str()
+            .chars()
+            .enumerate()
+            .skip_while(|(idx, _)| line_range.start > *idx)
+            .take(line_range.len())
+            .map(|(_, c)| c)
+            .collect::<String>();
+
+        let terminal_size = get_terminal_size();
+
+        // this values is dynamically calculated for each line where the doc is.
+        // the line being analysed can affect how the indentation is done.
+        let padding_till_excerpt_start = indent + 2;
+
+        // Check whether the statement is too long the terminal size for fitting purposes.
```

Comment:
this comment should be moved to inside the function I guess where the implicit `else` condition happens